cql3/statements/describe_statement.cc

```diff
@@ -160,14 +161,72 @@ description type(replica::database& db, const lw_shared_ptr<keyspace_metadata>&
     return description(db, *udt, true);
 }
 
-future<std::vector<description>> types(replica::database& db, const lw_shared_ptr<keyspace_metadata>& ks, bool with_stmt = false) {
-    auto user_types_map = ks->user_types().get_all_types();
-    auto user_types = boost::copy_range<std::vector<shared_ptr<const keyspace_element>>>(user_types_map | boost::adaptors::transformed([] (const auto& e) {
-        // return static_pointer_cast<const keyspace_element>(e.second);
-        return e.second;
-    }));
+struct udt_node {
+    user_type udt;
+    std::vector<sstring> used_by; 
+};
+
+static future<> udts_sorting_aux(const sstring& name,
+        const std::unordered_map<sstring, udt_node>& udt_map, 
+        std::unordered_map<sstring, bool>& visited_map, 
+        std::stack<user_type>& result) {
+    visited_map[name] = true;
+    for (auto& u: udt_map.at(name).used_by) {
+        if (!visited_map[u]) {
+            co_await udts_sorting_aux(u, udt_map, visited_map, result);
+            co_await coroutine::maybe_yield();
+        }
+    }
+    result.push(udt_map.at(name).udt);
+}
+
+// Sort UDTs in a topological order.
+// It's requires to safely restore schema from description, because one UDT can depend on another.
+// The function sorts topologically using DFS algorithm.
+static future<std::vector<shared_ptr<const keyspace_element>>> get_sorted_udts(const lw_shared_ptr<keyspace_metadata>& ks) {  
```

Comment:
I think it is possible to write this algorithm in a simpler way, without using recursion and using a bit less state.

```c++
std::map<sstring, size_t> udt_to_use_count_by_others; // For each UDT, how many other UDTs not yet pushed to `sorted_order` directly depend on it?
std::vector<sstring> sorted_order; // UDTs in reverse order to the desired one.
```

- Phase 1: populate `udt_to_use_count_by_others` by going over all UDTs, calculating the UDTs they directly depend on and incrementing the counters appropriately.
- Phase 2: go over all UDTs and push those to `sorted_order` for which `udt_to_use_count_by_others[udt] == 0`.
- Phase 3: iterate over UDTs in `sorted_order` (including those that you push during this phase):
    - Calculate the UDTs the current UDT depends on and _decrement_ their counters. If a counter goes down to 0, push that UDT to the end of `sorted_order`.

This will compute a topological ordering for a _reverse_ graph. If you then reverse the order of elements in the final `sorted_order`, you will get a valid result for our problem.